### PR TITLE
FIX: Allow user to define a custom block explorer URL #7077

### PR DESCRIPTION
**Pull request by marcosrdz**
## Release Notes

- **New Features**
	- Users can now select and manage their preferred block explorer for transaction viewing and settings.
	- Introduced a customizable settings component for entering custom block explorer URLs.

- **Enhancements**
	- The application now dynamically generates URLs for transactions based on user-selected block explorers, improving flexibility.
	- Added error handling and prompts for better user experience when managing block explorer settings.
	- New screen component for accessing block explorer settings has been added to the navigation stack.

- **Localization**
	- New text entries added to improve user guidance related to block explorer settings.

These updates enhance the configurability and usability of the application, providing a more tailored experience for users.

### CODE REVIEW

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

File: models/blockExplorer.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    return selectedExplorer || BLOCK_EXPLORERS.DEFAULT; // Return the selected explorer or default to mempool.space
+  } catch (error) {
+    console.error('Error getting block explorer:', error);
+    return BLOCK_EXPLORERS.DEFAULT;
+  }
+};
+
+export const saveBlockExplorer = async (url: string): Promise<boolean> => {
+  try {
+    await DefaultPreference.setName(GROUP_IO_BLUEWALLET);
```
Comment (Overtorment) [inline models/blockExplorer.ts:25]: need to validate url i recon (at least that it starts with http:// or https:// )

File: components/TransactionListItem.tsx [hunk 2]
```diff
@@ -253,16 +253,16 @@ export const TransactionListItem: React.FC<TransactionListItemProps> = React.mem
     const handleOnCopyTransactionID = useCallback(() => Clipboard.setString(item.hash), [item.hash]);
     const handleOnCopyNote = useCallback(() => Clipboard.setString(subtitle ?? ''), [subtitle]);
     const handleOnViewOnBlockExplorer = useCallback(() => {
-      const url = `https://mempool.space/tx/${item.hash}`;
+      const url = `${selectedBlockExplorer}/${item.hash}`;
```
Comment (Overtorment) [inline components/TransactionListItem.tsx:256]: `/tx/` should be here instring

File: components/TransactionListItem.tsx [hunk 3]
```diff
@@ -253,16 +253,16 @@ export const TransactionListItem: React.FC<TransactionListItemProps> = React.mem
...
         if (supported) {
           Linking.openURL(url);
         }
       });
-    }, [item.hash]);
+    }, [item.hash, selectedBlockExplorer]);
     const handleCopyOpenInBlockExplorerPress = useCallback(() => {
-      Clipboard.setString(`https://mempool.space/tx/${item.hash}`);
-    }, [item.hash]);
+      Clipboard.setString(`${selectedBlockExplorer}/${item.hash}`);
```
Comment (Overtorment) [inline components/TransactionListItem.tsx:264]: `/tx/` should be here instring

File: models/blockExplorer.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import DefaultPreference from 'react-native-default-preference';
+import { GROUP_IO_BLUEWALLET } from '../blue_modules/currency';
+
+export const BLOCK_EXPLORER = 'blockExplorer';
+export const BLOCK_EXPLORERS = {
+  DEFAULT: 'https://mempool.space/tx',
```
Comment (Overtorment) [inline models/blockExplorer.ts:6]: `/tx` should not be part of DEFAULT.
as well as other options

File: screen/settings/SettingsBlockExplorer.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  return (
+    <FlatList
+      data={blockExplorers}
+      keyExtractor={item => item.key}
+      renderItem={renderItem}
+      contentInsetAdjustmentBehavior="automatic"
+      automaticallyAdjustContentInsets
+      style={[styles.container, { backgroundColor: colors.background }]}
+    />
+  );
```
Comment (Overtorment) [inline screen/settings/SettingsBlockExplorer.tsx:81]: do you want to offer a selection of blockexplorers? ok

File: screen/settings/SettingsBlockExplorer.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  const { selectedBlockExplorer, setBlockExplorerStorage } = useSettings();
+  const [customUrlInput, setCustomUrlInput] = useState<string>(selectedBlockExplorer || '');
+  const { colors } = useTheme();
+
+  const blockExplorers = [
+    { name: `${loc._.default} - Mempool.space`, key: 'default', url: BLOCK_EXPLORERS.DEFAULT },
+    { name: 'Blockchair', key: 'blockchair', url: BLOCK_EXPLORERS.BLOCKCHAIR },
+    { name: 'Blockstream.info', key: 'blockstream', url: BLOCK_EXPLORERS.BLOCKSTREAM },
+    { name: loc.settings.block_explorer_custom, key: 'custom', url: null },
+  ];
```
Comment (Overtorment) [inline screen/settings/SettingsBlockExplorer.tsx:20]: its bad that your predefined explorers are spread across this file and  `models/blockExplorer.ts`

**Review by Overtorment — CHANGES_REQUESTED** [review]

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

File: models/blockExplorer.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
+// blockExplorer.ts
```
Comment (Overtorment) [inline models/blockExplorer.ts:1]: did chatgpt write this code?

**Review by Overtorment — APPROVED** [review]
lgtm

**Comment by GladosBlueWallet** [comment]
Unbelievable. You, [subject name here], must be the pride of [subject hometown here]!

**Pull request merged**